board: nucleo_f429zi: add partitions to support mcuboot

jli157 opened the pull request:
Add partition table to support MCUBoot.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 committed:
board: nucleo_f429zi: add partitions to support mcuboot

Add partition table to support MCUBoot. A paragraph of
usage comments was also added in the board documentation.

Signed-off-by: Jun Li <jun.r.li@intel.com>

erwango commented on boards/arm/nucleo_f429zi/nucleo_f429zi.dts:
```diff
@@ -96,3 +97,51 @@ arduino_serial: &usart6 {};
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/* 128K for bootloader */
```
Isn't it a bit large

jli157 replied:
Yes, but the first several sectors are like below:
```
 * sector 0     16K   0x000 0000
 * sector 1     16K
 * sector 2     16K
 * sector 3     16K
 * sector 4     64K
```

So, either using not more than 64K or 128K. 64K is a bit tight.

erwango replied:
ok

erwango commented on boards/arm/nucleo_f429zi/nucleo_f429zi.dts:
```diff
@@ -96,3 +97,51 @@ arduino_serial: &usart6 {};
+			reg = <0x00060000 0x00040000>;
+		};
+
+		/* swap slot: 256KB */
```
I thought mcuboot was using only a small slot to do the swap, is 256k really needed?

jli157 replied:
the scratch partition is used to temporarily store the entire data in either slot0 or slot1. So, its size should be the same as that of slot0 or slot1.

erwango replied:
According to https://github.com/JuulLabs-OSS/mcuboot/blob/master/docs/design.md#image-slots:
`The scratch area must have a size that is enough to store at least the largest sector that is going to be swapped`
And considering the image size, this leaves around ~5000 upgrade cycles, which is more than  enough.

erwango commented:
Partition is application dependent.
I'm not opposed to put a partition for this board, but maybe some explanations on the targetted application and choices made would be useful here.

jli157 commented:
> Partition is application dependent.
> I'm not opposed to put a partition for this board, but maybe some explanations on the targetted application and choices made would be useful here.

I agree. However, most board definitions in zephyr are used as references for application design. I guess most developers would like to define their boards based on their specific applications, like what I'm doing.:) So, having the partition table here will be a good reference.  A developer can copy the entire board directory here as his/her starting point and does some changes to adjust the partition size .

erwango commented:
> I agree. However, most board definitions in zephyr are used as references for application design. I guess most developers would like to define their boards based on their specific applications, like what I'm doing.:) So, having the partition table here will be a good reference. A developer can copy the entire board directory here as his/her starting point and does some changes to adjust the partition size .

I totally agree. Though, this board (like other nucleo), is quite generic and not tainted for a particular application (except maybe Ethernet, which is not very specific). 
Ideally provided partition would also be as 'generic' as the board... which  does not make much sense.
Hence, it is ok and welcome to set define a partition that could be used as example, even if targeting a specific application.
Though, in order to help others to tune partition for their needs from this example, it would be nice to have bit more comments on the way these sizes have been defined. Discussing these during the review is a good start, but I think a paragraph in board documentation would be a good achievement.
Could you work on that?

erwango commented on boards/arm/nucleo_f429zi/nucleo_f429zi.dts:
```diff
@@ -96,3 +97,51 @@ arduino_serial: &usart6 {};
+			reg = <0x000e0000 0x00020000>;
+		};
+
+		/* data partition: 1MB */
```
Can you elaborate on this choice as well?
Looking to the big picture, we have a 2MB flash size available and we end up with 256K image (which is admittedly already big for zephyr), this might be well worth a comment.

jli157 replied:
Is it possible to use macros in dts files so that partition size definitions can be changed based on a specific application?

jli157 commented:
> > I agree. However, most board definitions in zephyr are used as references for application design. I guess most developers would like to define their boards based on their specific applications, like what I'm doing.:) So, having the partition table here will be a good reference. A developer can copy the entire board directory here as his/her starting point and does some changes to adjust the partition size .
> 
> I totally agree. Though, this board (like other nucleo), is quite generic and not tainted for a particular application (except maybe Ethernet, which is not very specific).
> Ideally provided partition would also be as 'generic' as the board... which does not make much sense.
> Hence, it is ok and welcome to set define a partition that could be used as example, even if targeting a specific application.
> Though, in order to help others to tune partition for their needs from this example, it would be nice to have bit more comments on the way these sizes have been defined. Discussing these during the review is a good start, but I think a paragraph in board documentation would be a good achievement.
> Could you work on that?

Okay, I'll write comments in the board documentation about this as well. By the way, the partition table is actually useless if the macro `CONFIG_BOOTLOADER_MCUBOOT` is not enabled. The application can still use as much space as the flash permits.

jli157 commented:
@erwango reduced the sizes for bootloader and settings, removed the last partition. Just keep those really necessary for Zephyr + mcuboot. Also, I've added a paragraph in the documentation to introduce the partition table as a reference. 

Please help to check again. Thanks!

erwango commented on boards/arm/nucleo_f429zi/nucleo_f429zi.dts:
```diff
@@ -96,3 +97,46 @@ arduino_serial: &usart6 {};
+			reg = <0x00010000 0x00010000>;
+		};
+
+		slot0_partition: partition@20000 {
```
Can you set the comment back?

jli157 replied:
Done.

galak commented:
@dbkinder can you review the doc change.

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -182,6 +182,22 @@ Programming and Debugging
+The STM32F429ZI MCU on the board has an internal flash memory with 2MBytes in size.
+To use `MCUboot`_, a `zephyr partition table`_ for the flash memory needs to be defined
+in its device tree file ``nucleo_f429zi.dts``. As a reference, a partition table for MCUBoot
+is already defined in the device tree file, in which
```
```suggestion
is already defined in the device tree file, with these settings:
```

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -182,6 +182,22 @@ Programming and Debugging
+***************************************
+
+The STM32F429ZI MCU on the board has an internal flash memory with 2MBytes in size.
+To use `MCUboot`_, a `zephyr partition table`_ for the flash memory needs to be defined
```
```suggestion
To use `MCUboot`_, define a :ref:`Zephyr partition table <flash_partitions>`
for the flash memory
```

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -197,3 +213,9 @@ This interface is supported by the openocd version included in Zephyr SDK.
 .. _STM32F429 datasheet:
    http://www.st.com/resource/en/datasheet/DM00071990.pdf
+
+.. _zephyr partition table:
```
This link target doesn't exist.  I suggested a change above to use a ``:ref:`` to link to the correct document location, so you can remove these two lines (217-218)

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -182,6 +182,22 @@ Programming and Debugging
+is already defined in the device tree file, in which
+
+- `MCUBoot`_ bootloader partition takes 64K bytes.
+- zephyr settings partition takes 64K bytes.
```
```suggestion
- Zephyr settings partition takes 64K bytes.
```

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -182,6 +182,22 @@ Programming and Debugging
+Flash partitions for MCUBoot bootloader
+***************************************
+
+The STM32F429ZI MCU on the board has an internal flash memory with 2MBytes in size.
```
```suggestion
The on-board STM32F429ZI MCU has 2 MBs of internal flash memory.
```

dbkinder reviewed (changes requested):
Some clarity edits and fixes for a broken link.

jli157 commented:
> Some clarity edits and fixes for a broken link.

Hi David, I've corrected the broken link to the proper one, and applied other changes you had suggested. Can you review again? Thank you!

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -182,6 +182,22 @@ Programming and Debugging
+***************************************
+
+The on-board STM32F429ZI MCU has 2MBs of internal flash memory. To use `MCUboot`_,
+define a `Zephyr partition table`_ for the flash memory in its device tree file
```
While what you're linking to now will work, it always points to the /latest/ documentation. When we create a tagged version of the documentation, we want references to be to content within that tagged version. So the right way to write this link is:
```
define a :ref:`Zephyr partition table <flash_partitions>` for the flash memory
```
The Sphinx document processor will create a link within the document tree it's building instead of always referencing the /latest/ version.

jli157 replied:
Thanks for pointing out the potential problem, David! I've already corrected it again by following your comments.

dbkinder commented on boards/arm/nucleo_f429zi/doc/index.rst:
```diff
@@ -197,3 +213,9 @@ This interface is supported by the openocd version included in Zephyr SDK.
+.. _MCUBoot:
+   https://github.com/JuulLabs-OSS/mcuboot/blob/master/README.md
+
+.. _Zephyr partition table:
```
use a :ref: role (as noted above) and remove this link (lines 220-221)

jli157 commented:
@dbkinder can you recheck the changes? Thank you!

dbkinder reviewed (approved):
doc changes LGTM, thanks!
